File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool isPossible(vector<int>& weights, int days, int capacity){
        int sum = 0;
        int numdays = 1;

        for(int i = 0; i < weights.size(); i++){
            if(sum + weights[i] <= capacity){
                sum += weights[i];
            }else{
                numdays++;
                sum = weights[i];
            }
        }

        if(numdays <= days){
            return true;
        }
        return false;
    }

    int shipWithinDays(vector<int>& weights, int days) {
        int low = *max_element(weights.begin(), weights.end());
        int high = accumulate(weights.begin(), weights.end(), 0);
        int ans = 0;

        while(low <= high){
            int mid = (low + high) / 2;

            if(isPossible(weights, days, mid) == true){
                ans = mid;
                high = mid - 1;
            }else{
                low = mid + 1;
            }
        }
        return ans;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.py
class Solution:
    def isPossible(self, weights, capacity, days):
        numdays = 1
        summ = 0
        for i in range(len(weights)):
            if summ + weights[i] <= capacity:
                summ += weights[i]
            else:
                numdays += 1
                summ = weights[i]

        if numdays <= days:
            return True
        
        return False

    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low = max(weights)
        high = sum(weights)

        ans = 0

        while low <= high:
            mid = (low + high) // 2

            if self.isPossible(weights, mid, days) == True:
                ans = mid
                high = mid - 1
            else:
                low = mid + 1
        
        return ans
